### ISSUE TITLE
Fast-cache key (stringify_obj_type) omits gradient state for array args

### ISSUE THREAD

**Issue by hughperkins** [Genesis-Embodied-AI/quadrants#882]
## Summary

`python/quadrants/lang/_fast_caching/args_hasher.py::stringify_obj_type()` produces a persistent fast-cache type string that omits gradient state for every array argument type:

- `torch.Tensor` -> `[pt-{dtype}-{ndim}]`
- `numpy.ndarray` -> `[np-{dtype}-{ndim}]`
- `ScalarNdarray` (`qd.ndarray`) -> `[nd-{dtype}-{ndim}]` (and `[ndv-...]`, `[ndm-...]`)

The in-memory `TemplateMapper` *does* specialize on gradient state (`_extract_arg` folds `needs_grad` / `requires_grad` into the spec key), so two args that differ only in gradient state produce distinct in-memory specializations but collide to the same persistent fast-cache key.

## Potential impact

When a fast-cache-eligible kernel is first cached with grad-state A and later invoked in another runtime / process with the same dtype and rank but grad-state B, the persistent cache can serve frontend code compiled for grad-state A, so the gradient buffer may not be bound as expected -> potentially incorrect autodiff.

Note: this tracks the *key collision*. Whether it produces observable autodiff corruption end-to-end (vs. being re-bound at launch, or fastcache being disabled for grad kernels) still needs an end-to-end cross-process repro.

## Evidence (CPU, quadrants `33359ecd8`)

`stringify_obj_type()` output - `requires_grad` False vs True collide, identically for `qd.Tensor` and `qd.types.NDArray` slots; native ndarray `needs_grad` False vs True also collide:

```
[qd.Tensor slot        ] requires_grad=False -> '[pt-torch.float32-1]'
[qd.Tensor slot        ] requires_grad=True  -> '[pt-torch.float32-1]'   collide=True
[qd.types.NDArray slot ] requires_grad=False -> '[pt-torch.float32-1]'
[qd.types.NDArray slot ] requires_grad=True  -> '[pt-torch.float32-1]'   collide=True
ndarray needs_grad=False -> '[nd-f32-1]'
ndarray needs_grad=True  -> '[nd-f32-1]'   collide=True
```

## Scope / notes

Pre-existing and global; not specific to any annotation. Surfaced during codex review of #878 (which widens `qd.Tensor` to accept external arrays), but #878 does not introduce it - a torch tensor at a `qd.types.ndarray()` slot hashes identically, and quadrants-native ndarrays share the omission.

A fix should:
- add gradient state to the persistent repr for all array reprs (`pt` / `nd` / `ndv` / `ndm`; `np` has no grad concept), mirroring the exact `needs_grad` derivation used by `_extract_arg` (annotation override, else `requires_grad` / allocation flag);
- account for the offline-cache invalidation this key-format change causes for existing array-taking kernels;
- ideally land with an end-to-end cross-process autodiff regression test.
